art.go

```diff
@@ -20,79 +24,131 @@ package art
 
 // TODO: IPv4 + IPv6 constructors
 
-func baseIndex(width int, addr uint64, prefixLen int) uint64 {
-	return (addr >> uint64(width-prefixLen)) + (1 << uint64(prefixLen))
+func baseIndex(width int, addr uint32, prefixLen int) uint32 {
+	return (addr >> uint32(width-prefixLen)) | (1 << uint32(prefixLen))
 }
 
-func fringeIndex(width int, addr uint64) uint64 {
+func fringeIndex(width int, addr uint32) uint32 {
 	return baseIndex(width, addr, width)
 }
 
+func getBits(byteOffset, bits int, src []byte) uint32 {
+	if byteOffset > len(src) {
+		return 0
+	}
+	var v uint32
+	src = src[:(len(src) - byteOffset)]
+	switch len(src) {
+	case 0:
+		return 0
+	case 1:
+		v = uint32(src[0])
+	case 2:
+		v = uint32(src[1]) + (uint32(src[0]) << 8)
+	case 3:
+		v = uint32(src[2]) +
+			(uint32(src[1]) << 8) +
+			(uint32(src[0]) << 16)
+	default:
+		v = uint32(src[len(src)-4]) +
+			(uint32(src[len(src)-3]) << 8) +
+			(uint32(src[len(src)-2]) << 16) +
+			(uint32(src[len(src)-1]) << 24)
+	}
+	return v & ((1 << bits) - 1)
+}
+
 // A Route is an entry in the routing table.
 type Route interface {
 	// RouteParams returns the properties of the routing table entry.
 	RouteParams() RouteParams
+	Equals(Route) bool
 }
 
 // RouteParams are the properties of the routing table entry.
 type RouteParams struct {
-	Width int    // bits of routable porition address (32-bit for IPv4, 64-bit for IPv6)
-	Addr  uint64 // the destination address
-	Len   int    // prefix length of route (e.g. 24 for a 192.168.0.0/24)
+	Addr []byte // the destination address
+	Len  int    // prefix length of route (e.g. 24 for a 192.168.0.0/24)
```

Comment:
then again, maybe RouteParams can just go away and you can use netaddr.IPPrefix instead?